src/server/worker/common/common.go

```diff
@@ -23,16 +23,12 @@ func IsDone(ctx context.Context) bool {
 
 // DatumID computes the ID of a datum.
 func DatumID(inputs []*Input) string {
-	// TODO: This is a stopgap solution that needs to be addressed before 2.0 GA.
-	// Datum IDs need more design work.
-	if inputs[0].GroupBy != "" {
-		return inputs[0].GroupBy
-	}
-	var files []string
+	hash := pfs.NewHash()
 	for _, input := range inputs {
-		files = append(files, input.Name+strings.ReplaceAll(input.FileInfo.File.Path, "/", "_"))
+		hash.Write([]byte(input.Name))
```

Comment:
I think these need to be prefixed with their length to prevent collisions. Use the binary package to serialize the length as an integer and just write it to the hash.

Reply:
I'm not sure if I am following the suggestion here, but I think we could handle the duplicate input / file name pair issue by including the index of the input in the hash. This may be what you are suggesting, but I'm not sure based on how you worded it. This would double down on the order being relevant, but I think that is a reasonable constraint to have. I guess we could do some pre-sorting of the inputs to make the IDs stable regardless of the input order.

Reply:
If two paths to concatenate to another path that is the problem.  The index could work, I think there just has to be additional bytes the input doesn't have control over.  I think this is called a "padding scheme" if you want to google for it.  I'm not sure if it helps to have the extra bytes be derived from (e.g. the length) the input.

Reply:
Look at TupleHash
https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-185.pdf

Reply:
I see. Yeah, I think the index approach should address both problems.

Reply:
I'm not sure about the index. I think you could still have a problem in some cases when the index is in the printable range.

I found a Go implementation of TupleHash and I think we should just copy and paste the encodeString, and leftEncode, rightEncode functions from there.
https://github.com/chain/txvm/blob/ff6bfbe5389214685247420092f4fe1befbf3a2e/crypto/sha3/cshake.go#L77

Reply:
We discussed this offline. Duplicate datums (same IDs) will be filtered out, so the only problem to address is the collisions due to concatenations. We are addressing this by writing the length of each write (input name, path, input name, path, ...) into the hasher.